# Amplicons sequencing workbook

- this is a workspace for Ajoke and Steve to interact and keep track of work undertaken to analyse amplicon sequencing data



## What we want to achieve
- Aim: to optimise the amplicon sequencing approach
- What are we working with: Amplicon sequencing data generated by Umer Chaudrhy
- What we want to discover:

1. How many worms have sequencing data?
    - How reads per worm?
    - Step1 - multiplex PCR per worm
    - Step2 - pool the PCR products from each worm into a single tube
        - RESULT - how good were the worms pooled? How much DNA was present
    - Q. do we need to measure the concentration of DNA to pool
    - Q. is there enough DNA to produce PCR products

2. Of the worms that have sequencing data, how many reads map to the genome
    - We “expected” that all the reads map to the genome. Do they?
        - Eg. could get primer dimers that produce sequencing reads, but don’t map
    - What proportion of all reads map?

3. Of the reads that map to the genome, what proportion of the reads map to where we expect them to map.
    - Primers have been designed to amplify specific regions
    - What proportion of reads map correctly to where we designed the primers?

4. Of the reads that map correctly, how many reads do we get per amplicon
    - We hope that all amplicons are present, but we dont know this
    - Are all amplicons present?
    - Are the present in equal proportions?
    - Have any clearly not worked, or worked too well?



---
### 1. How many worms have sequencing data?

```bash
# counting lines using word count or "wc"
wc -l *fastq
```
- this prints all the information, however, it still needs some interpretation. 
- how about we do something slightly more sophisticated. Just a little. Lets use a "for loop"

```bash
# basic loop, and we'll print the name
for NAME in *_1.fastq; do
    echo -e "${NAME}";
    done
```
- try this. It should print a list of the file names


- if this works, lets add some extra commands to calculate the number of reads
```bash
for NAME in *_1.fastq; do 
    # count the lines, and extract the 1st column using "awk" and divide it by 4
    READS=$( wc -l ${NAME} | awk '{print $1/4}' ) ;
    # print the name and read count
    echo -e "${NAME}\t${READS}" >> raw_reads.count ;     
    done 

# create new file
> new.file

# add to an existing file
>> new.file
```bash
# count the number of reads
for NAME in *1.fastq; do READS=$( wc -1)${NAME} | awk {print $1/4} ); echo -e "${NAME}\t${READS}" >> raw_reads.count; done
# fastqc
for NAME in *fastq; do fastqc $NAME; done
# multiqc
multiqc ..
firefox multiqc_report.html
Total number of reads for ALL18 Dataset = 3506725 with 76 samples. The total number of reads expected per 




# mapping reads using a loop
for NAME in $( ls -1 *_1.fastq | sed 's/_1.fastq//g'); do
    bwa mem haemonchus_contortus.PRJEB506.WBPS15.genomic.fa.gz ${NAME}_1.fastq ${NAME}_2.fastq > ${NAME}.mapping.sam ;
    
    done





# getting amplicon coordinates 
# working dir: /nfs/users/nfs_s/sd21/lustre118_link/hc/AMPLICONS/primers

cat primers.Hc_XQTL_BZ_chr1_5-10Mb.10000bp_windows_200000bp_apart.database | cut -f3,6,2 | sed -e 's/:/\t/g' -e 's/-/\t/g' -e 's/_P/\tP/g' | awk '{print $2,$3,$8,$1}' OFS="\t" | grep -v "primer"

#> bed file - chromosome, start, end, name

hcontortus_chr1_Celeg_TT_arrow_pilon	5007667	5007972	PCR1
hcontortus_chr1_Celeg_TT_arrow_pilon	5206863	5207162	PCR2
hcontortus_chr1_Celeg_TT_arrow_pilon	5407809	5408126	PCR3
hcontortus_chr1_Celeg_TT_arrow_pilon	5606291	5606592	PCR4
hcontortus_chr1_Celeg_TT_arrow_pilon	5807329	5807610	PCR5
hcontortus_chr1_Celeg_TT_arrow_pilon	6007701	6007985	PCR6
hcontortus_chr1_Celeg_TT_arrow_pilon	6208711	6209018	PCR7
hcontortus_chr1_Celeg_TT_arrow_pilon	6402081	6402349	PCR8
hcontortus_chr1_Celeg_TT_arrow_pilon	6806063	6806361	PCR9
hcontortus_chr1_Celeg_TT_arrow_pilon	7007057	7007329	PCR10
hcontortus_chr1_Celeg_TT_arrow_pilon	7204321	7204609	PCR11
hcontortus_chr1_Celeg_TT_arrow_pilon	7401460	7401770	PCR12
hcontortus_chr1_Celeg_TT_arrow_pilon	7604940	7605213	PCR13
hcontortus_chr1_Celeg_TT_arrow_pilon	7809354	7809643	PCR14
hcontortus_chr1_Celeg_TT_arrow_pilon	8003534	8003806	PCR15
hcontortus_chr1_Celeg_TT_arrow_pilon	8207374	8207682	PCR16
hcontortus_chr1_Celeg_TT_arrow_pilon	8403840	8404137	PCR17
hcontortus_chr1_Celeg_TT_arrow_pilon	8603538	8603843	PCR18
hcontortus_chr1_Celeg_TT_arrow_pilon	8809661	8809953	PCR19
hcontortus_chr1_Celeg_TT_arrow_pilon	9004304	9004619	PCR20
hcontortus_chr1_Celeg_TT_arrow_pilon	9203534	9203804	PCR21
hcontortus_chr1_Celeg_TT_arrow_pilon	9400550	9400843	PCR22
hcontortus_chr1_Celeg_TT_arrow_pilon	9603081	9603352	PCR23
hcontortus_chr1_Celeg_TT_arrow_pilon	9801439	9801745	PCR24


# extract reads that map to amplicons 

samtools view -b -L btub.amplicons.bed XQTL_F3_L3_n5k_IVM_pre_01_23204_8_1.merged.sorted.marked.realigned.bam -o test.bam
samtools flagstat test.bam



# coverage of amplicons 

bedtools multicov -bams XQTL_F3_L3_n5k_IVM_pre_01_23204_8_1.merged.sorted.marked.realigned.bam -bed btub.amplicons.bed





# SNP calling
ls -1 *bam > bams.list

bcftools mpileup --ignore-RG -Ou --min-MQ 20 --adjust-MQ 50 --bam-list bams.list --fasta-ref HAEM_V4_final.chr.fa --skip-indels -E --regions-file regions.bed | bcftools call -vm -Oz -o variants.vcf.gz
